fish-rust/src/builtins/string/match.rs

```diff
@@ -0,0 +1,410 @@
+
+        if let StringMatcher::Regex(RegexMatcher {
+            first_match_captures,
+            ..
+        }) = matcher
+        {
+            let vars = parser.get_vars();
+            for (name, vals) in first_match_captures.into_iter() {
+                vars.set(&WString::from(name), EnvMode::DEFAULT, vals);
+            }
+        }
+
+        if match_count > 0 {
+            STATUS_CMD_OK
+        } else {
+            STATUS_CMD_ERROR
+        }
+    }
+}
+
+struct RegexMatcher<'opts, 'args> {
+    regex: Box<Regex>,
+    total_matched: usize,
+    first_match_captures: HashMap<String, Vec<WString>>,
+    opts: &'opts Match<'args>,
+}
+
+struct WildCardMatcher<'opts, 'args> {
+    pattern: WString,
+    total_matched: usize,
+    opts: &'opts Match<'args>,
+}
+
+enum StringMatcher<'opts, 'args> {
+    Regex(RegexMatcher<'opts, 'args>),
+    WildCard(WildCardMatcher<'opts, 'args>),
+}
+
+impl<'opts, 'args> StringMatcher<'opts, 'args> {
+    fn new(
+        pattern: &'args wstr,
+        opts: &'opts Match<'args>,
+    ) -> Result<StringMatcher<'opts, 'args>, RegexError> {
+        if opts.regex {
+            let m = RegexMatcher::new(pattern, opts)?;
+            Ok(Self::Regex(m))
+        } else {
+            let m = WildCardMatcher::new(pattern, opts);
+            return Ok(Self::WildCard(m));
+        }
+    }
+
+    fn report_matches(
+        &mut self,
+        arg: &wstr,
+        streams: &mut io_streams_t,
+    ) -> Result<(), pcre2::Error> {
+        match self {
+            Self::Regex(m) => m.report_matches(arg, streams)?,
+            Self::WildCard(m) => m.report_matches(arg, streams),
+        }
+        Ok(())
+    }
+
+    fn match_count(&self) -> usize {
+        match self {
+            Self::Regex(m) => m.total_matched,
+            Self::WildCard(m) => m.total_matched,
+        }
+    }
+}
+
+enum MatchResult<'a> {
+    NoMatch,
+    Match(Option<Captures<'a>>),
+}
+
+impl<'opts, 'args> RegexMatcher<'opts, 'args> {
+    fn new(
+        pattern: &'args wstr,
+        opts: &'opts Match<'args>,
+    ) -> Result<RegexMatcher<'opts, 'args>, RegexError> {
+        let regex = RegexBuilder::new()
+            .caseless(opts.ignore_case)
+            .never_utf(true)
+            .build(pattern.as_char_slice())
+            .map_err(|e| RegexError::Compile(pattern.to_owned(), e))?;
+
+        Self::validate_capture_group_names(regex.capture_names())?;
+
+        let first_match_captures = regex
+            .capture_names()
+            .iter()
+            .filter_map(|name| name.as_ref().map(|n| (n.to_owned(), Vec::new())))
+            .collect();
+        let m = Self {
+            regex: Box::new(regex),
+            total_matched: 0,
+            first_match_captures,
+            opts,
+        };
+        return Ok(m);
+    }
+
+    fn report_matches(
+        &mut self,
+        arg: &wstr,
+        streams: &mut io_streams_t,
+    ) -> Result<(), pcre2::Error> {
```

Comment:
I'm reviewing commit-by-commit; it threw me off that the first commit did some weird stuff that was changed later

Reply:
Actually all commits except the second and the last one should be squashed I think.
I'll do that locally and continue reviewing

Reply:
I agree the changes were weird, and I'm very fine with squashing